sycl-fusion/passes/internalization/Internalization.cpp

```diff
@@ -290,6 +280,43 @@ Error SYCLInternalizerImpl::canPromoteCall(CallBase *C, const Value *Val,
   return Error::success();
 }
 
+enum GEPKind { INVALID = 0, NEEDS_REMAPPING, ADDRESSES_INTO_AGGREGATE };
+
+static int getGEPKind(GetElementPtrInst *GEPI, const PromotionInfo &PromInfo) {
+  assert(GEPI->getNumIndices() >= 1 && "No-op GEP encountered");
+
+  // Inspect the GEP's source element type.
+  auto &DL = GEPI->getModule()->getDataLayout();
+  auto SrcElemTySz = DL.getTypeAllocSize(GEPI->getSourceElementType());
+
+  // `GEPI`'s first index is selecting elements. Unless it is constant zero, we
+  // have to remap. If there are more indices, we start to address into an
+  // aggregate type.
+  if (SrcElemTySz == PromInfo.ElemSize) {
+    int Kind = INVALID;
+    if (!match(GEPI->idx_begin()->get(), m_ZeroInt()))
+      Kind |= NEEDS_REMAPPING;
+    if (GEPI->getNumIndices() >= 2)
+      Kind |= ADDRESSES_INTO_AGGREGATE;
+    assert(Kind != INVALID && "No-op GEP encountered");
+    return Kind;
+  }
+
+  // Check whether `GEPI` adds a constant offset, e.g. a byte offset to address
+  // into a padded structure, smaller than the accessor's element size.
+  MapVector<Value *, APInt> VariableOffsets;
+  auto PW = DL.getPointerSizeInBits(GEPI->getPointerAddressSpace());
+  APInt ConstantOffset = APInt::getZero(PW);
+  if (GEPI->collectOffset(DL, PW, VariableOffsets, ConstantOffset) &&
```

Comment:
Why are we using `PW` instead of `PromInfo.ElemSize` there?

Reply:
The `collectOffset` method accumulates offsets internally in the bitwidth of pointers in the GEP's address space, hence it's a different thing than our buffer's element size.

Reply:
I see. I'm not sure it makes much difference, but [here](https://github.com/intel/llvm/blob/ba44e2e31672a3fcf3d4d0615f359ad822f74c30/llvm/lib/IR/Operator.cpp#L184) they're asserting that argument is equal to `DL.getIndexSizeInBits(getPointerAddressSpace())`. LGTM after this is changed if needed.

Reply:
Good catch!